crates/holochain_sqlite/src/db.rs

```diff
@@ -201,43 +219,113 @@ pub trait WriteManager<'e> {
+impl<'e> WriteManager<'e> for DbWrite {
+    fn with_commit<E, R, F>(&'e mut self, f: F) -> Result<R, E>
     where
-        F: 'e + FnOnce(Transaction) -> R,
+        E: From<DatabaseError> + std::fmt::Debug,
+        F: 'e + Clone + FnOnce(&mut Transaction) -> Result<R, E>,
     {
-        self.with_reader(|r| DatabaseResult::Ok(f(r))).unwrap()
+        self.conn()?.with_commit(f)
     }
 }
 
-impl<'e> WriteManager<'e> for PConn {
-    fn with_commit<E, R, F>(&'e mut self, f: F) -> Result<R, E>
+impl<'e> ReadManager<'e> for PConn {
+    #[tracing::instrument(skip(self, f))]
+    fn with_reader<E, R, F>(&'e mut self, f: F) -> Result<R, E>
     where
         E: From<DatabaseError>,
-        F: 'e + FnOnce(&mut Transaction) -> Result<R, E>,
+        F: 'e + FnOnce(Transaction) -> Result<R, E>,
     {
-        let mut txn = self
-            .transaction_with_behavior(TransactionBehavior::Exclusive)
+        tracing::trace!("Opened DB read transaction.");
+        let txn = self
+            .transaction_with_behavior(TransactionBehavior::Deferred)
             .map_err(DatabaseError::from)?;
-        let result = f(&mut txn)?;
-        txn.commit().map_err(DatabaseError::from)?;
-        Ok(result)
+        let result = f(txn);
+        tracing::trace!("Closed DB read transaction.");
+        result
+    }
+}
+
+impl<'e> WriteManager<'e> for PConn {
+    #[tracing::instrument(skip(self, f))]
+    fn with_commit<E, R, F>(&'e mut self, f: F) -> Result<R, E>
+    where
+        E: From<DatabaseError> + std::fmt::Debug,
+        F: 'e + Clone + FnOnce(&mut Transaction) -> Result<R, E>,
+    {
+        let backoff = backoff::ExponentialBackoff {
+            initial_interval: Duration::from_millis(5),
+            multiplier: 1.5,
+            randomization_factor: 0.1,
+            // TODO: reduce these timeouts significantly after doing some initial
+            // testing to find out how long of a timeout, if any,
+            // fixes our database locking issues.
+            max_interval: Duration::from_secs(1),
+            max_elapsed_time: Some(Duration::from_secs(5)),
+            ..Default::default()
+        };
+
+        // NB: The `?` here converts all errors into `backoff::Error::Transient`.
+        //     If there are unrecoverable errors, we can explicitly map those
+        //     to `backoff::Error::Permanent`.
+        let attempt = || {
+            tracing::trace!(
+                "Attempting to acquire lock on txn mutex for {:?}.",
+                self.db_kind()
+            );
+            let mx = get_txn_mutex(self.db_kind());
+            let _g = mx.lock();
+            tracing::trace!("Acquired lock on txn mutex for {:?}.", self.db_kind());
+
+            let mut txn = self
+                .transaction_with_behavior(TransactionBehavior::Exclusive)
+                .map_err(DatabaseError::from)
+                .map_err(E::from)?;
+            let result = (f.clone())(&mut txn)?;
```

Comment:
I'm not sure how big the data in the closure might be but seems this strategy forces a clone even if there's no contention. Might be a bit of a heavy lift if you're writing large items like a dna to the db?

Reply:
I definitely want to remove the retry (and clones) if the mutex fixes the problem. But I think it's better to add as many remedies as possible since we our testing time with Holo is precious. The retries are just a big ol' safety net for this next test.

I still maintain that it's possible that there are some edge cases in sqlite that we aren't aware of where it returns Busy immediately, and the retries will show us if that's the case, but the busy_timeout won't. The retries will also show us if we just need to up the busy_timeout (though 30s should be more than enough, if we're doing everything else right).

As for the clones, I'm hoping this is the kind of thing rustc can optimize away, similar to how `Cow` is sort of irrelevant a lot of the time due to rustc optimizing away such clones. But again we'll remove that if no retries are actually needed.

Reply:
That's fair although you could achieve the same thing by setting the busy_timeout to 5 hours right? Because if you see any busy errors then you know that it's not from the busy timeout and it's some edge case that you are referring too? 
This is not your fault in anyway but it's a shame we have to change our code to suit a slow testing framework. Definitely something worth bringing up at the next stand up we are all at.